File: Lab/PracticalAssignment7.cpp
#include <iostream>
#define Max 20
#define infi 999
using namespace std;
class Graph
{
private:
    int A[Max][Max];
    int NoOfvertices;

public:
    Graph()
    {
        cout << "ENTER NO OF VERTICES: " << endl;
        cin >> NoOfvertices;
        for (int i = 0; i < NoOfvertices; i++)
        {
            for (int j = 0; j < NoOfvertices; j++)
            {
                if (i != j)
                {
                    A[i][j] = infi;
                }
                else
                {
                    A[i][j] = 0;
                }
            }
        }
    }
    void Create_Graph()
    {
        char r;
        int w;
        for (int i = 0; i < NoOfvertices; i++)
        {
            for (int j = 0; j < NoOfvertices; j++)
            {
                if (i == j)
                {
                    continue;
                }
                else
                {

                    cout << "WHETHER " << i << " & " << j << " ARE ADJACENT TO EACH OTHER: Y/N?";
                    cin >> r;
                    if ((r == 'y') || (r == 'Y'))
                    {
                        cout << "ENTER THE WEIGHT: " << endl;
                        cin >> A[i][j];
                    }
                }
            }
        }
    }
    void Display_Graph()
    {
        cout << "YOUR GRAPH IS AS FOLLOWS: " << endl;
        for (int i = 0; i < NoOfvertices; i++)
        {
            for (int j = 0; j < NoOfvertices; j++)
            {
                cout << A[i][j] << "\t";
            }
            cout << endl;
        }
    }
    void display_MST()
    {
        int visited[NoOfvertices], sv, min, x, y;
        for (int i = 0; i < NoOfvertices; i++)
        {
            visited[i] = 0;
        }
        cout << "ENTER STARTING VERTEX: " << endl;
        cin >> sv;
        visited[sv] = 1;
        int no_edge = 0;
        int cost_MST = 0;
        cout << "EDGE: WEIGHT " << endl;
        while (no_edge < NoOfvertices - 1)
        {
            min = infi;
            x = 0;
            y = 0;
            for (int i = 0; i < NoOfvertices; i++)
            {
                if (visited[i])
                {
                    for (int j = 0; j < NoOfvertices; j++)
                    {
                        if (!visited[j] && A[i][j])
                        {
                            if (min > A[i][j])
                            {
                                min = A[i][j];
                                x = i;
                                y = j;
                            }
                        }
                    }
                }
            }
            cout << x << "-->" << y << " :" << A[x][y] << endl;
            cost_MST += A[x][y];
            visited[y] = 1;
            no_edge++;
        }
        cout << "COST OF MST IS: " << cost_MST << endl;
    }
    void Total_Cost()
    {
        int tc = 0;
        for (int i = 0; i < NoOfvertices; i++)
        {
            for (int j = 0; j < NoOfvertices; j++)
            {
                if (i < j && (A[i][j]!=999||A[i][j]!=infi))
                {
                    tc += A[i][j];
                }
            }
        }
        cout << "TOTAL COST IS: " << tc << endl;
    }
};

int main()
{
    Graph g;
    int n;
    cout << "1.CREATE GRAPH " << endl;
    cout << "2.DISPLAY GRAPH " << endl;
    cout << "3.DISPLAY OF MINIMUM SPANNING TREE " << endl;
    cout << "4.TO DISPLAY TOTAL COST " << endl;
    do
    {

        cout << "\nENTER YOUR CHOICE: " << endl;
        cin >> n;
        switch (n)
        {
        case 1:
            g.Create_Graph();
            break;
        case 2:
            g.Display_Graph();
            break;
        case 3:
            g.display_MST();
            break;
        case 4:
            g.Total_Cost();
            break;
        default:
            break;
        }

    } while (n != 5);
}

File: Lab/PracticalAssignment7.py
import sys

MAX = 20
INF = 999

class Graph:
    def __init__(self, n):
        self.n = n
        self.A = [[INF if i != j else 0 for j in range(n)] for i in range(n)]

    def create_graph(self):
        for i in range(self.n):
            for j in range(self.n):
                if i != j:
                    r = input(f"Are {i} and {j} adjacent? (Y/N)").lower()
                    if r == 'y':
                        w = int(input(f"Enter the weight of the edge ({i}, {j}): "))
                        self.A[i][j] = w


    def display_graph(self):
        print("Your graph is:")
        for row in self.A:
            print("\t".join(str(cell) for cell in row))

    def display_mst(self):
        visited = [0] * self.n
        print("Enter the starting vertex:")
        sv = int(input())
        visited[sv] = 1
        no_edge = 0
        cost_mst = 0
        print("Edges: Weight")
        while no_edge < self.n - 1:
            m = INF
            x = y = 0
            for i in range(self.n):
                if visited[i]:
                    for j in range(self.n):
                        if not visited[j] and self.A[i][j]:
                            if m > self.A[i][j]:
                                m = self.A[i][j]
                                x = i
                                y = j
            print(f"{x} -> {y}: {self.A[x][y]}")
            cost_mst += self.A[x][y]
            visited[y] = 1
            no_edge += 1
        print("Cost of MST:", cost_mst)

    def total_cost(self):
        tc = sum(self.A[i][j] for i in range(self.n) for j in range(i+1, self.n) if self.A[i][j])
        print("Total cost:", tc)


n = 0
while n != 5:
    print("1. Create graph")
    print("2. Display graph")
    print("3. Display minimum spanning tree")
    print("4. Display total cost")
    print("5. Exit")
    n = int(input("Enter your choice: "))
    
    if n == 1:
        print("Enter the number of vertices:")
        no_of_vertices = int(input())
        g = Graph(no_of_vertices)
        g.create_graph()
    elif n == 2:
        g.display_graph()
    elif n == 3:
        g.display_mst()
    elif n == 4:
        g.total_cost()
    elif n == 5:
        break
    else:
        print("Invalid choice.")
